File: backend/src/tax_calc.py
"""Shared GST line math — shelf/line amounts are always tax-inclusive.

Discounts (line-item and document-level) are applied to the inclusive
amount first; GST is then extracted from what remains.
"""

from __future__ import annotations

# Kept for callers that still pass a mode; exclusive is no longer supported.
VALID_TAX_PRICING_MODES = frozenset({"inclusive"})


def normalize_tax_pricing_mode(mode: str | None) -> str:
    """Always inclusive — org preference removed."""
    return "inclusive"


def line_tax_amount(line_amount: float, tax_rate: float, mode: str | None = None) -> float:
    """Tax component extracted from a GST-inclusive discounted line amount."""
    if tax_rate <= 0 or line_amount <= 0:
        return 0.0
    return round(line_amount * tax_rate / (100 + tax_rate), 2)


def line_taxable_amount(line_amount: float, tax_rate: float, mode: str | None = None) -> float:
    """Pre-tax (taxable) portion of an inclusive line amount."""
    tax = line_tax_amount(line_amount, tax_rate, mode)
    return round(line_amount - tax, 2)


def allocate_flat_amount(amounts: list[float], flat: float) -> list[float]:
    """Split a document-level discount across inclusive line amounts.

    Proportional by line size; the last positive line absorbs remainder so
    shares sum to ``min(flat, sum(amounts))``.
    """
    n = len(amounts)
    if n == 0:
        return []
    disc = round(max(0.0, float(flat or 0)), 2)
    total = round(sum(max(0.0, float(a or 0)) for a in amounts), 2)
    if disc <= 0 or total <= 0:
        return [0.0] * n
    disc = min(disc, total)
    shares = [0.0] * n
    last_idx = 0
    for i, raw in enumerate(amounts):
        if float(raw or 0) > 0:
            last_idx = i
    allocated = 0.0
    for i, raw in enumerate(amounts):
        if i == last_idx:
            continue
        amount = max(0.0, float(raw or 0))
        if amount <= 0:
            continue
        share = min(amount, round(amount / total * disc, 2))
        shares[i] = share
        allocated += share
    remainder = round(disc - allocated, 2)
    last_cap = max(0.0, float(amounts[last_idx] or 0))
    shares[last_idx] = min(last_cap, max(0.0, remainder))
    return shares


def rollup_inclusive_lines(
    line_inclusives: list[float],
    tax_rates: list[float],
    entity_discount: float = 0.0,
) -> tuple[list[tuple[float, float, float]], float, float, float]:
    """Apply document discount, then extract GST from remaining inclusive amounts.

    Line-item discounts must already be reflected in ``line_inclusives``.

    Returns:
        rows: (inclusive_after_entity, taxable, tax) per input line
        subtotal: sum of taxable (after all discounts)
        tax_total: sum of tax (after all discounts)
        total: subtotal + tax_total (= payable inclusive amount)
    """
    shares = allocate_flat_amount(list(line_inclusives), entity_discount)
    rows: list[tuple[float, float, float]] = []
    subtotal = 0.0
    tax_total = 0.0
    for amount, rate, share in zip(line_inclusives, tax_rates, shares):
        after = round(max(0.0, float(amount or 0) - share), 2)
        tax = line_tax_amount(after, rate or 0)
        taxable = line_taxable_amount(after, rate or 0)
        rows.append((after, taxable, tax))
        subtotal += taxable
        tax_total += tax
    subtotal = round(subtotal, 2)
    tax_total = round(tax_total, 2)
    return rows, subtotal, tax_total, round(subtotal + tax_total, 2)

File: frontend/src/utils/taxCalc.js
import { formatAmountInput, roundAmount } from '@/utils/decimalPrecision'

/** Always inclusive — pricing-mode preference removed. */
export function normalizeTaxPricingMode(_mode) {
  return 'inclusive'
}

export function exclusiveFromInclusive(inclusive, taxRate) {
  const amount = Number(inclusive) || 0
  const rate = Number(taxRate) || 0
  if (amount <= 0) return 0
  if (rate <= 0) return roundAmount(amount)
  return roundAmount((amount * 100) / (100 + rate))
}

export function inclusiveFromExclusive(exclusive, taxRate) {
  const amount = Number(exclusive) || 0
  const rate = Number(taxRate) || 0
  if (amount <= 0) return 0
  if (rate <= 0) return roundAmount(amount)
  return roundAmount(amount * (1 + rate / 100))
}

/** Split an entered unit price into excl. GST / GST / incl. GST. */
export function priceTaxBreakdown(entered, mode, taxRate) {
  const n = Math.max(0, Number(entered) || 0)
  const inclusive = mode === 'exclusive'
    ? inclusiveFromExclusive(n, taxRate)
    : roundAmount(n)
  const exclusive = exclusiveFromInclusive(inclusive, taxRate)
  return {
    exclusive,
    tax: roundAmount(inclusive - exclusive),
    inclusive,
  }
}

/** Catalog amounts are always stored GST-inclusive. */
export function catalogInclusiveAmount(entered, mode, taxRate) {
  if (entered === '' || entered == null) return null
  const n = Number(entered)
  if (!Number.isFinite(n)) return null
  return priceTaxBreakdown(n, mode === 'exclusive' ? 'exclusive' : 'inclusive', taxRate).inclusive
}

/** Re-express a stored GST-inclusive catalog amount in another entry mode (for display). */
export function convertEnteredTaxAmount(value, fromMode, toMode, taxRate) {
  if (value === '' || value == null) return value
  const n = Number(value)
  if (!Number.isFinite(n) || n < 0) return value
  const from = fromMode === 'exclusive' ? 'exclusive' : 'inclusive'
  const to = toMode === 'exclusive' ? 'exclusive' : 'inclusive'
  if (from === to) return value
  const { inclusive } = priceTaxBreakdown(n, from, taxRate)
  if (to === 'inclusive') return formatAmountInput(inclusive)
  return formatAmountInput(exclusiveFromInclusive(inclusive, taxRate))
}

/** Tax amount extracted from one GST-inclusive discounted line total. */
export function lineTaxAmount(lineTotal, taxRate, _mode) {
  const amount = Number(lineTotal) || 0
  const rate = Number(taxRate) || 0
  if (amount <= 0 || rate <= 0) return 0
  return roundAmount(amount * rate / (100 + rate))
}

/** Split a document-level discount across inclusive line amounts. */
export function allocateFlatShares(amounts, flat) {
  const values = (amounts || []).map((a) => Math.max(0, Number(a) || 0))
  const n = values.length
  if (n === 0) return []
  const total = roundAmount(values.reduce((s, a) => s + a, 0))
  let disc = roundAmount(Math.max(0, Number(flat) || 0))
  if (disc <= 0 || total <= 0) return values.map(() => 0)
  disc = Math.min(disc, total)
  const shares = values.map(() => 0)
  let lastIdx = 0
  values.forEach((amount, i) => {
    if (amount > 0) lastIdx = i
  })
  let allocated = 0
  values.forEach((amount, i) => {
    if (i === lastIdx || amount <= 0) return
    const share = Math.min(amount, roundAmount((amount / total) * disc))
    shares[i] = share
    allocated += share
  })
  const remainder = roundAmount(disc - allocated)
  shares[lastIdx] = Math.min(values[lastIdx], Math.max(0, remainder))
  return shares
}

/**
 * Cart totals for POS checkout display and sale completion.
 * Line totals already include line-item discount. Bill discount is applied
 * next; GST is extracted from the remaining inclusive amount.
 * @param {Array<{ lineTotal: number, taxRate: number }>} cart
 */
export function calcCartTotals(cart, { discountPct = 0, discountAmt = 0 } = {}) {
  const lineTotals = (cart || []).map((item) => Number(item.lineTotal) || 0)
  const rates = (cart || []).map((item) => Number(item.taxRate) || 0)
  const grossSubtotal = roundAmount(lineTotals.reduce((s, n) => s + n, 0))

  const discount = roundAmount(
    (Number(discountAmt) || 0) + grossSubtotal * ((Number(discountPct) || 0) / 100),
  )
  const shares = allocateFlatShares(lineTotals, discount)

  let taxTotal = 0
  let netSubtotal = 0
  lineTotals.forEach((lineTotal, i) => {
    const after = roundAmount(Math.max(0, lineTotal - (shares[i] || 0)))
    const tax = lineTaxAmount(after, rates[i])
    taxTotal += tax
    netSubtotal += roundAmount(after - tax)
  })

  return {
    mode: 'inclusive',
    netSubtotal: roundAmount(netSubtotal),
    grossSubtotal,
    taxTotal: roundAmount(taxTotal),
    discount,
    total: Math.max(0, roundAmount(grossSubtotal - discount)),
  }
}

/**
 * Invoice summary math for printed tax invoices.
 * Gross amount is the raw pre-discount calculation based on qty × price.
 * Discount is the amount stripped from that gross before GST is added back.
 */
export function calcInvoiceSummary(items = [], sale = {}) {
  const grossAmount = roundAmount(
    (items || []).reduce((sum, item) => {
      const qty = Number(item?.qty ?? item?.quantity ?? 0)
      const price = Number(item?.price ?? item?.rate ?? 0)
      return sum + (qty * price)
    }, 0),
  )

  const lineNetAmount = roundAmount(
    (items || []).reduce((sum, item) => {
      const qty = Number(item?.qty ?? item?.quantity ?? 0)
      const price = Number(item?.price ?? item?.rate ?? 0)
      const lineTotal = Number(item?.lineTotal ?? item?.total ?? qty * price)
      return sum + lineTotal
    }, 0),
  )
  const headerDiscount = roundAmount(Number(sale?.discount ?? sale?.discount_amount ?? sale?.discount_amt ?? 0))
  const subtotal = roundAmount(Number(sale?.subtotal ?? sale?.netSubtotal ?? grossAmount))
  const taxTotal = roundAmount(Number(sale?.taxTotal ?? sale?.tax_total ?? 0))
  const total = roundAmount(Number(sale?.total ?? subtotal + taxTotal))
  const discountAmount = roundAmount(Math.max(0, grossAmount - lineNetAmount) + headerDiscount)

  return {
    grossAmount,
    subtotal,
    taxTotal,
    discountAmount,
    total,
  }
}
